storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageClient.java

```diff
@@ -21,54 +21,109 @@
 import net.consensys.cava.bytes.Bytes;
 import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
 import tech.pegasys.artemis.datastructures.operations.Attestation;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
 
 /** This class is the ChainStorage client-side logic */
 public class ChainStorageClient implements ChainStorage {
 
-  protected LinkedBlockingQueue<BeaconBlock> unprocessedBlocks;
-  protected HashMap<Bytes, Bytes> processedBlocks;
-  protected LinkedBlockingQueue<Attestation> unprocessedAttestations;
+  protected final LinkedBlockingQueue<BeaconBlock> unprocessedBlocks = new LinkedBlockingQueue<>();
+  protected final LinkedBlockingQueue<Attestation> unprocessedAttestations =
+      new LinkedBlockingQueue<>();
+  protected final HashMap<Bytes, BeaconBlock> processedBlockLookup = new HashMap<>();
+  protected final HashMap<Bytes, BeaconState> stateLookup = new HashMap<>();
   protected EventBus eventBus;
 
   public ChainStorageClient() {}
 
   public ChainStorageClient(EventBus eventBus) {
-    this.unprocessedBlocks = new LinkedBlockingQueue<BeaconBlock>();
-    this.unprocessedAttestations = new LinkedBlockingQueue<Attestation>();
-    this.processedBlocks = new HashMap<Bytes, Bytes>();
     this.eventBus = eventBus;
     this.eventBus.register(this);
   }
 
-  @Override
-  public void addProcessedBlock(Bytes blockHash, BeaconBlock block) {
-    ChainStorage.<Bytes, HashMap<Bytes, Bytes>>add(blockHash, block.toBytes(), processedBlocks);
-    LOG.info("BlockHash: " + blockHash.toHexString() + " added to ChainStorage.");
+  /**
+   * Add processed block to storage
+   *
+   * @param blockHash
+   * @param block
+   */
+  public void addProcessedBlock(Bytes state_root, BeaconBlock block) {
+    ChainStorage.<Bytes, BeaconBlock, HashMap<Bytes, BeaconBlock>>add(
+        state_root, block, this.processedBlockLookup);
+    LOG.info(
+        "Block with state root: " + state_root.toHexString() + " has been added to ChainStorage.");
     // todo: post event to eventbus to notify the server that a new processed block has been added
   }
 
-  @Override
+  /**
+   * Add calculated state to storage
+   *
+   * @param state_root
+   * @param state
+   */
+  public void addState(Bytes state_root, BeaconState state) {
+    ChainStorage.<Bytes, BeaconState, HashMap<Bytes, BeaconState>>add(
+        state_root, state, this.stateLookup);
+    LOG.info(
+        "Block with state root: " + state_root.toHexString() + " has been added to ChainStorage.");
```

Comment:
Might wanna Log something different than processing a block here. I initially got confused thinking this is another method trying to store blocks.

Reply:
good catch. done